mmgen/models/architectures/biggan/biggan_snmodule.py

```diff
@@ -0,0 +1,257 @@
+'''
+# yapf:enable
+
+
+def proj(x, y):
+    """Calculate Projection of x onto y.
+
+    Args:
+        x (torch.Tensor): Projection vector x.
+        y (torch.Tensor): Direction vector y.
+
+    Returns:
+        torch.Tensor: Projection of x onto y.
+    """
+    return torch.mm(y, x.t()) * y / torch.mm(y, y.t())
+
+
+def gram_schmidt(x, ys):
+    """Orthogonalize x wrt list of vectors ys.
+
+    Args:
+        x (torch.Tensor): Vector to be added into the
+            orthogonal vectors.
+        ys (list[torch.Tensor]): A set of orthogonal
+            vectors.
+
+    Returns:
+        torch.Tensor: Result of Gram–Schmidt
+            orthogonalization.
+    """
+    for y in ys:
+        x = x - proj(x, y)
+    return x
+
+
+def power_iteration(W, u_, update=True, eps=1e-12):
+    """Power iteration method for calculating spectral norm.
+
+    Args:
+        W (torch.Tensor): Module weight.
+        u_ (list[torch.Tensor]): list of left singular vector.
+            The length of list equals to the simulation times.
+        update (bool, optional): Whether update left singular
+            vector. Defaults to True.
+        eps (float, optional): Vector Normalization epsilon.
+            Defaults to 1e-12.
+
+    Returns:
+        tuple[list[tensor.Tensor]]: Tuple consist of three lists
+            which contain singular values, left singular
+            vector and right singular vector respectively.
+    """
+    us, vs, svs = [], [], []
+    for i, u in enumerate(u_):
+        with torch.no_grad():
+            v = torch.matmul(u, W)
+            v = F.normalize(gram_schmidt(v, vs), eps=eps)
+            vs += [v]
+            u = torch.matmul(v, W.t())
+            u = F.normalize(gram_schmidt(u, us), eps=eps)
+            us += [u]
+            if update:
+                u_[i][:] = u
+        svs += [torch.squeeze(torch.matmul(torch.matmul(v, W.t()), u.t()))]
+    return svs, us, vs
+
+
+class SN(object):
+    """Spectral normalization base class.
+
+    Args:
+        num_svs (int): Number of singular values.
+        num_itrs (int): Number of power iterations per step.
+        num_outputs (int): Number of output channels.
+        transpose (bool, optional): If set to `True`, weight
+            matrix will be transposed before power iteration.
+            Defaults to False.
+        eps (float, optional): Vector Normalization epsilon for
+            avoiding divide by zero. Defaults to 1e-12.
+    """
+
+    def __init__(self,
+                 num_svs,
+                 num_itrs,
+                 num_outputs,
+                 transpose=False,
+                 eps=1e-12):
+        self.num_itrs = num_itrs
+        self.num_svs = num_svs
+        self.transpose = transpose
+        self.eps = eps
+        # Register a singular vector for each sv
+        for i in range(self.num_svs):
+            self.register_buffer('u%d' % i, torch.randn(1, num_outputs))
```

Comment:
It is confusing why you do not inherit from `nn.Module`, but you can call `self.register_buffer`.